dataplicity/m2m/wsclient.py

```diff
@@ -199,7 +199,27 @@ def closed(self, code, reason=None):
         try:
             self.on_close(self)
         except Exception as error:
-            log.error('error on WSApp.closed: %s', error)
+            log.error('WSApp.closed: %s', error)
+
+    def unhandled_error(self, error):
+        """Called by ws4py when there is an error in run_forever."""
+        # ws4py calls this with any socket errors
+        # Doesn't matter what the socket error is; connection is fubar.
+        log.error('error in WSApp: %s', error)
+        try:
+            self.terminate()
+        except Exception as terminate_error:
+            log.error('WSApp.unhandled_error: %s', terminate_error)
+
+    def close_connection(self):
```

Comment:
If the heartbeat thread fails, it does **not** call `unhandled_error` above -- it just calls `close_connection` and exits.

Since there is no official hook, I call on_close from here, so the client knows to re-connect.